out_kafka: Add dynamic/static headers support

DIFRIN opened the pull request:
What

  Add a new header configuration option to the out_kafka plugin that
  attaches custom Kafka message headers to every produced message.

  Two header modes are supported:

  - Static — the configured value is used verbatim.
  - Dynamic — a value prefixed with $ is resolved at runtime from the
  matching log record field. If the field does not exist or is not a
  string, a warning is emitted and that header is silently skipped; the
  message is still produced.

  Memory ownership

  rd_kafka_producev() with RD_KAFKA_V_HEADERS() takes ownership of
  the rd_kafka_headers_t object on success only. The plugin calls
  rd_kafka_headers_destroy() on every failure path:

  - Non–queue-full produce error
  - Queue-full retry limit exceeded (queue_full_retries)

  ---
  Testing

  

- [x]  Example configuration file for the change

  [INPUT]
      Name    dummy
      Tag     app.logs
      Dummy   {"level": "info", "host": "web-01", "trace_id": "abc-123"}

  [OUTPUT]
      Name        kafka
      Match       *
      Brokers     192.168.1.3:9092
      Topics      app-events

      # Static headers — value used verbatim
      header      X-Environment  production
      header      X-Source       fluent-bit

      # Dynamic headers — value resolved from the log record field
      header      X-Host         $host
      header      X-TraceId      $trace_id

  YAML equivalent:

  pipeline:
    inputs:
      - name: dummy
        tag: app.logs
        dummy: '{"level":"info","host":"web-01","trace_id":"abc-123"}'

    outputs:
      - name: kafka
        match: '*'
        brokers: 192.168.1.3:9092
        topics: app-events
        header: 'X-Environment production'
        header: 'X-Source fluent-bit'
        header: 'X-Host $host'
        header: 'X-TraceId $trace_id'

  ---
  

- [x] Debug log output

  All five runtime tests passed. Key lines annotated below.

  Test raw_format...
  [2026/03/01 07:44:47] [ info] [output:kafka:kafka.0] brokers='127.0.0.1:111' topics='test'
  [2026/03/01 07:44:47] [error] [output:kafka:kafka.0] 127.0.0.1:111/bootstrap:
      Connect to ipv4#127.0.0.1:111 failed: Connection refused
  [2026/03/01 07:44:54] [ warn] [output:kafka:kafka.0] Failed to force flush:
      Local: Timed out
  [ OK ]

  Test header_static...              <-- two static headers: "env production", "team platform"
  [2026/03/01 07:44:54] [ info] [output:kafka:kafka.0] brokers='127.0.0.1:111' topics='test'
  [2026/03/01 07:44:54] [error] [output:kafka:kafka.0] 127.0.0.1:111/bootstrap:
      Connect to ipv4#127.0.0.1:111 failed: Connection refused
  [2026/03/01 07:45:02] [ warn] [output:kafka:kafka.0] Failed to force flush:
      Local: Timed out
  [ OK ]

  Test header_dynamic...             <-- "source $key_0" resolved to "val_0"
  [2026/03/01 07:45:02] [ info] [output:kafka:kafka.0] brokers='127.0.0.1:111' topics='test'
  [2026/03/01 07:45:02] [error] [output:kafka:kafka.0] 127.0.0.1:111/bootstrap:
      Connect to ipv4#127.0.0.1:111 failed: Connection refused
  [2026/03/01 07:45:10] [ warn] [output:kafka:kafka.0] Failed to force flush:
      Local: Timed out
  [ OK ]

  Test header_dynamic_missing_field... <-- "$no_such_field" absent: warn + skip
  [2026/03/01 07:45:10] [ info] [output:kafka:kafka.0] brokers='127.0.0.1:111' topics='test'
  [2026/03/01 07:45:10] [error] [output:kafka:kafka.0] 127.0.0.1:111/bootstrap:
      Connect to ipv4#127.0.0.1:111 failed: Connection refused
  [2026/03/01 07:45:11] [ warn] [output:kafka:kafka.0] header 'missing': field
      '$no_such_field' not found in record or not a string, skipping  <-- expected
  [2026/03/01 07:45:18] [ warn] [output:kafka:kafka.0] Failed to force flush:
      Local: Timed out
  [ OK ]

  Test header_mixed...               <-- "app myapp" (static) + "source $key_1" (dynamic)
  [2026/03/01 07:45:18] [ info] [output:kafka:kafka.0] brokers='127.0.0.1:111' topics='test'
  [2026/03/01 07:45:18] [error] [output:kafka:kafka.0] 127.0.0.1:111/bootstrap:
      Connect to ipv4#127.0.0.1:111 failed: Connection refused
  [2026/03/01 07:45:26] [ warn] [output:kafka:kafka.0] Failed to force flush:
      Local: Timed out
  [ OK ]

  SUCCESS: All unit tests have passed.

  ---
  

- [x] Valgrind output

  Command:
  valgrind --tool=memcheck --leak-check=full \
    --show-leak-kinds=definite,indirect \
    --error-exitcode=1 \
    --suppressions=valgrind.supp \
    build/bin/flb-rt-out_kafka

  Output:
  ==26539== Memcheck, a memory error detector
  ==26539== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
  ==26539== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
  ==26539== Command: build/bin/flb-rt-out_kafka

  Test raw_format...                              [ OK ]
  Test header_static...                           [ OK ]
  Test header_dynamic...                          [ OK ]
  Test header_dynamic_missing_field...            [ OK ]
  Test header_mixed...                            [ OK ]
  SUCCESS: All unit tests have passed.

  ==26539== HEAP SUMMARY:
  ==26539==     in use at exit: 0 bytes in 0 blocks
  ==26539==   total heap usage: 9,995 allocs, 9,995 frees, 6,314,321 bytes allocated
  ==26539==
  ==26539== All heap blocks were freed -- no leaks are possible
  ==26539==
  ==26539== For lists of detected and suppressed errors, rerun with: -s
  ==26539== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

  ---
 

- [x]  Documentation

  - Documentation updated in [fluent-bit-docs:](https://github.com/fluent/fluent-bit-docs/pull/1341) added header row to
  the parameter table and a new "Message headers" section with static +
  dynamic examples in both .conf and .yaml formats.

  ---
 

- [N/A ]  Backporting

  this is a new feature targeting master only; backport
  decisions deferred to maintainers.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Kafka output plugin supports custom message headers on produced messages; headers may be static or resolved dynamically from message fields using $field_name syntax.

* **Reliability**
  * Improved header lifecycle and error handling to prevent header leaks and ensure correct behavior across retry and failure scenarios.

* **Tests**
  * Added tests for static, dynamic, mixed headers and missing-field handling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DIFRIN commented:
@patrick-stephens DCO added to the commits

DIFRIN commented:
@patrick-stephens this is the message of the failed check 

`The following tests FAILED:
	 46 - flb-rt-out_http (Failed)
Error: Process completed with exit code 8.`

it's not linked to the PR content

patrick-stephens commented:
Possibly the test failure is related to https://github.com/fluent/fluent-bit/pull/9023

patrick-stephens commented:
> @patrick-stephens Hello Pat my team and I are strongly interested in using this new feature. Why is it still on hold ? Thank you for your answer  

You can see here that it is failing a couple of checks so cannot be merged until it succeeds. It likely needs rebasing first and then updating as required. 

I would encourage you to pick up the pr potentially and contribute the updates if you require it.

DIFRIN committed:
out_kafka: add static and dynamic Kafka message header support

Add a new 'header' configuration option to the Kafka output plugin,
allowing users to attach custom headers to every produced message.

Header values starting with '$' are treated as dynamic references: the
named field is looked up in the log record and its string value is used
as the header value. Any other value is used verbatim (static header).

Example:

  header X-Env      production
  header X-TraceId  $trace_id

Multiple 'header' entries may be provided. If a referenced field is not
found in the record or its value is not a string, the header is skipped
and a warning is logged.

The produce call was migrated from rd_kafka_produce() to the variadic
rd_kafka_producev(), which is required to pass the headers object via
RD_KAFKA_V_HEADERS(). On produce failure, rdkafka does not take
ownership of the headers object, so it is freed explicitly; on
queue-full retries the same object is reused safely.

Signed-off-by: difrin <frindi.abderrahmane.ied@gmail.com>

DIFRIN committed:
[Tests] out_kafka: add static and dynamic header test cases

 Add four runtime test cases covering the message header feature added
  in out_kafka. All tests use an unreachable broker (127.0.0.1:111)
  with queue_full_retries=1; no real Kafka instance is required.

  - header_static: two verbatim static headers; exercises the static
    branch of kafka_build_message_headers()
  - header_dynamic: value prefixed '$' resolved from the log record;
    exercises the dynamic field-lookup branch
  - header_dynamic_missing_field: referenced field absent from the
    record; verifies the warning-and-skip path is clean under Valgrind
  - header_mixed: one static + one dynamic header in the same message;
    exercises both branches together

  Tested with Valgrind 3.22.0: 0 errors, 0 leaks.

Signed-off-by: difrin <frindi.abderrahmane.ied@gmail.com>

DIFRIN commented on tests/runtime/out_kafka.c:
```diff
@@ -7,31 +7,201 @@
+    flb_output_set(ctx, out_ffd, "raw_log_key",        "key_0",         NULL);
+    flb_output_set(ctx, out_ffd, "topics",             "test",          NULL);
+    flb_output_set(ctx, out_ffd, "brokers",            "127.0.0.1:111", NULL);
+    flb_output_set(ctx, out_ffd, "queue_full_retries", "1",             NULL);
```
done

DIFRIN committed:
[Feature] out_kafka: Handle dynamic numeric types headers

 - Added char num_buf[64]; int num_len; locals before the loop
  - Split the old single if (found && is_string) / else warn into six branches:
    a. key not found → warn "not found in record, skipping"
    b. MSGPACK_OBJECT_STR → verbatim (unchanged)
    c. MSGPACK_OBJECT_POSITIVE_INTEGER → snprintf with PRIu64
    d. MSGPACK_OBJECT_NEGATIVE_INTEGER → snprintf with PRId64
    e. MSGPACK_OBJECT_FLOAT32 / FLOAT64 → snprintf with %g
    f. anything else (bool/array/map/nil) → warn "unsupported type, skipping"

 - Addes Helper for tests + updated block comments to mention integer/float conversion and the updated warning wording

Signed-off-by: difrin <frindi.abderrahmane.ied@gmail.com>

kimonus commented:
I have been working on the map-based variant requested in #1673 and discussed
again in #8334/#9448. Before opening another PR, I would like to confirm the
preferred API direction with the maintainers and this PR's author.

This PR currently proposes repeated `header` entries, with either static values
or values resolved from individual `$field` references. The alternative
implementation uses:

```yaml
headers_key: headers
```

The selected top-level record field is a map containing the complete set of
Kafka message headers. A valid map is omitted from the serialized payload. The
implementation currently preserves ordered and duplicate header names and
supports string, binary, empty, and null values. Missing fields are ignored,
while a selected non-map field is retained in the payload to avoid data loss.
`otlp_json` and `otlp_proto` reject the option during startup.

The map-based implementation also includes a Kafka magic-v2 record-batch parser
in the in-tree integration fake broker, so tests assert the headers that are
actually encoded on the wire rather than only checking plugin startup. Current
local evidence is:

- 8 focused dynamic-header integration cases passed.
- The same 8 cases passed with strict Valgrind checks.
- The complete out_kafka integration file passed (34 cases).
- The repository Dockerfile debug build passed.

The two approaches modify the same `produce_message()` /
`rd_kafka_producev()` ownership and retry path, but they serve different
configuration use cases. Would maintainers prefer that `headers_key`:

1. complement this PR's `header` option;
2. be submitted as a separate follow-up after this PR;
3. replace the per-header interface; or
4. be integrated into a clean successor to this PR?

I will hold off on publishing a competing source or documentation PR until the
preferred direction is clear. I can provide the implementation and integration
test patch for review once that direction is confirmed.

kimonus commented:
Hi @DIFRIN, following up directly because I did not mention you in my previous comment. Do you plan to continue working on this PR, including rebasing and updating it against the current `master`?

I am preparing a complementary `out_kafka` feature: `headers_key` reads the complete set of per-message Kafka headers from a record map. Your repeated `header` option handles individually configured static or dynamic headers, while `headers_key` handles a variable set supplied by each record.

The implementations touch the same `produce_message()` and `rd_kafka_producev()` path, so knowing your plans would help us avoid conflicting work and decide whether the map-based feature should be a follow-up or coordinated with this PR.
